### PR TITLE
Add a Registration example

### PR DESCRIPTION
**Pull request by ericmann**
In addition to the more complete CRUD on data example, add in a dynamic client registration illustration that both demonstrates how to create a new keypair and client as well as the reasoning behind doing so.

Fixes E3DB-686

### CODE REVIEW

File: examples/registration.rb [hunk 1]
```diff
@@ -0,0 +1,73 @@
+# This program provides a simple example illustrating how to programmatically
+# register a client with InnoVault and e3db. In some situations, it's preferable
+# to register a client from the server or system that will be using its
+# credentials (to ensure that all data is truly encrypted from end-to-end
+# with no possibilities of a credential leak). For more detailed information,
```
Comment (gstro) [inline examples/registration.rb]: > In some situations, it's preferable to register a client...

This section makes me think that credential leaks just happen in other flows. I'd reword or omit.

File: examples/registration.rb [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+# Clients can either create new cryptographic keypairs, or load in a pre-defined
+# pair of Curve25519 keys. In this situation, we will generate a new keypair.
+private_key, public_key = E3DB::Client.generate_keypair
+
+puts("Public Key:  " + public_key)
+puts("Private Key: " + private_key)
+
+# The e3db server keeps track of the name of the curve used with public keys,
+# so we need to wrap the generated version with an object helper
+wrapped_key = E3DB::PublicKey.new(:curve25519 => public_key)
```
Comment (gstro) [inline examples/registration.rb]: It seems like registration should handle this wrapping? I don't think we're aiming for flexibility (and shouldn't). And we don't support other curves atm, anyway.

File: examples/registration.rb [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+# Clients must be registered with a name unique to your account to help
+# differentiate between different sets of credentials in the Admin Console.
+# In this example, the name is set at random
+client_name = sprintf("client_%s", SecureRandom.hex)
+
+puts("Client Name: " + client_name)
+
+# Passing all of the data above into the registration routine will create
+# a new client with the system. Remember to keep your private key private!
```
Comment (gstro) [inline examples/registration.rb]: Nice.

File: examples/registration.rb [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+require 'e3db'
+
+# A registration token is required to set up a client. In this situation,
+# we assume an environment variable called REGISTRATION_TOKEN is set
+token = ENV["REGISTRATION_TOKEN"]
+
+# Clients can either create new cryptographic keypairs, or load in a pre-defined
+# pair of Curve25519 keys. In this situation, we will generate a new keypair.
+private_key, public_key = E3DB::Client.generate_keypair
```
Comment (ericmann) [inline examples/registration.rb]: The tuple of keys is (public, private), not (private, public). Need to fix ...
